internal/app.go

```diff
@@ -294,15 +314,36 @@ func (a *App) checkin(client *http.Client, crypto CryptoHandler) error {
 	if err != nil {
 		return fmt.Errorf("Unable to get: %s - %v", a.configUrl, err)
 	}
+	defer res.Body.Close()
+
 	if res.StatusCode == 200 {
+		body, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("Unable to read new secrets: %w", err)
+		}
+
+		var config configSnapshot
+		if config.next, err = UnmarshallBuffer(crypto, body, true); err != nil {
+			return err
+		}
+		if config.prev, err = UnmarshallFile(nil, a.EncryptedConfig, false); err != nil {
+			log.Printf("Unable to load previous config version: %s", err)
+			var perr *os.PathError
+			if !errors.As(err, &perr) || !os.IsNotExist(perr) {
+				return err
+			}
+		}
+
+		if err = a.extract(crypto, config); err != nil {
+			return err
+		}
+
 		modtime, err := time.Parse(time.RFC1123, res.Header.Get("Date"))
 		if err != nil {
 			log.Printf("Unable to get modtime of config file, defaulting to 'now': %s", err)
 			modtime = time.Now()
 		}
-		if err := safeWrite(res.Body, a.EncryptedConfig, modtime); err != nil {
-			return err
-		}
+		return safeWrite(body, a.EncryptedConfig, modtime)
```

Comment:
This change in behavior (switch an order of saving a new config file and applying changes) has one important side effect described below.

Assume that there is a problem extracting files on disk (e.g. disk is full or file encryption is broken or whatever).
Old behavior:
- Store a new config version to disk, try to extract it, fail and forget until a new version is available.
New behavior:
- Try to extract a new config version, fail and at the next check-in try again until success.
- This try..fail..try-again loop may continue almost endlessly until a problem preventing a file extraction if fixed.

@doanac @mike-scott @ricardosalveti do you think this is a desired behavior ?
If not - I can forcefully save a new file at the end, even if extraction fails.

Reply:
@doanac should confirm, but I think the new behavior is better and more predictable from the user perspective.